javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -79,1587 +91,989 @@ class Formatter(
-          formatToken.getStrFlags(),
-          formatToken.getConversionType())
-      if (formatToken.isPrecisionSet())
-        throw new IllegalFormatPrecisionException(formatToken.getPrecision())
-      padding(result, startIndex)
+  private def parsePositiveIntSilent(capture: Option[String],
+                                     default: Int): Int = {
+    capture.fold {
+      default
+    } { s =>
+      val x = JDouble.parseDouble(s)
+      if (x <= Int.MaxValue) x.toInt
+      else -1 // Silently ignore and return -1
     }
+  }
 
-    private def transformFromLineSeparator(): String = {
-      if (formatToken.isPrecisionSet())
-        throw new IllegalFormatPrecisionException(formatToken.getPrecision())
-
-      if (formatToken.isWidthSet())
-        throw new IllegalFormatWidthException(formatToken.getWidth())
-
-      val flags = formatToken.getFlags()
-      if (FormatToken.FLAGS_UNSET != flags)
-        throw new IllegalFormatFlagsException(formatToken.getStrFlags())
-
-      if (null == lineSeparator) {
-        lineSeparator = System.getProperty("line.separator")
+  private def formatArg(localeInfo: LocaleInfo,
+                        arg: Any,
+                        conversion: Char,
+                        flags: Flags,
+                        width: Int,
+                        precision: Int): Unit = {
+    @inline def rejectPrecision(): Unit = {
+      if (precision >= 0)
+        throw new IllegalFormatPrecisionException(precision)
+    }
+
+    def formatNullOrThrowIllegalFormatConversion(): Unit = {
+      if (arg == null)
+        formatNonNumericString(localeInfo, flags, width, precision, "null")
+      else {
+        throw new IllegalFormatConversionException(conversion, arg.getClass)
       }
-      lineSeparator
     }
 
-    private def padding(source: StringBuilder, startIndex: Int): String = {
-      var start: Int        = startIndex
-      val paddingRight      = formatToken.isFlagSet(FormatToken.FLAG_MINUS)
-      var paddingChar: Char = '\u0020' // space as padding char.
-      if (formatToken.isFlagSet(FormatToken.FLAG_ZERO)) {
-        if ('d' == formatToken.getConversionType())
-          paddingChar = getDecimalFormatSymbols().getZeroDigit()
-        else
-          paddingChar = '0'
-      } else {
-        start = 0
-      }
-      var width     = formatToken.getWidth()
-      val precision = formatToken.getPrecision()
+    @inline def precisionWithDefault =
+      if (precision >= 0) precision
+      else 6
 
-      var length = source.length()
-      if (precision >= 0) {
-        length = Math.min(length, precision)
-        source.delete(length, source.length())
-      }
-      if (width > 0) {
-        width = Math.max(source.length(), width)
+    @inline def efgCommon(notation: (Double, Int, Boolean) => String): Unit = {
+      val arg0 = arg match {
+        case arg: Double     => arg
+        case arg: Float      => arg.toDouble
+        case arg: BigDecimal => arg.doubleValue()
```

Comment:
This is not correct, as it will drop a lot of precision from the `BigDecimal`. We'll need to address this.